### PR TITLE
More usage options

### PR DESCRIPTION
**Pull request by kablamo**
```
use MooX::Options 
   usage_top    => ['USAGE_TOP'],    # adds text after usage string but before the options text
   usage_bottom => ['USAGE_BOTTOM'], # adds text after the options text
   man_option   => 0,  # remove --man option
   help_option  => 0,  # remove --help option
   usage_option => 0;  # remove --usage option
```

See the included test for an example.

### CODE REVIEW

File: lib/MooX/Options/Descriptive/Usage.pm [hunk 1]
```diff
@@ -134,6 +134,10 @@ sub option_text {
     my @message;
     my $lf = _get_line_fold();
     for my $opt(@$getopt_options) {
+        if ($opt->{desc} eq 'spacer') {
```
Comment (celogeek) [inline lib/MooX/Options/Descriptive/Usage.pm]: What does that mean ?

File: lib/MooX/Options/Role.pm [hunk 2]
```diff
@@ -278,13 +278,13 @@ sub parse_options {
...
     my $usage_str = $options_config{usage_string} // "USAGE: $prog_name %o";
 
-    my ( $opt, $usage ) = describe_options(
-        ($usage_str), @$options,
-        [ 'usage', 'show a short help message'],
-        [ 'help|h', "show a help message" ],
-        [ 'man', "show the manual" ],
-        ,@flavour
-    );
+    unshift @$options, $options_config{usage_top}          if $options_config{usage_top};
```
Comment (celogeek) [inline lib/MooX/Options/Role.pm]: What is this ? is it an option ?
To push it on top, use "order" params. order => -1 = on top, order => 999 = on bottom.

If you want to add test before the usage, and after, that should be another kind of patch.

File: lib/MooX/Options/Role.pm [hunk 3]
```diff
@@ -314,18 +314,15 @@ sub parse_options {
         }
     }
 
-    if (   $opt->help() || defined $params{help}
```
Comment (celogeek) [inline lib/MooX/Options/Role.pm]: the attributes is defined in the role, but I don't see patch that remove them when you don't want them. Also you may have forget the check on existence of those methods if option is disabled.

File: lib/MooX/Options/Role.pm (hunk 2 above)
Comment (kablamo) [inline lib/MooX/Options/Role.pm]: It is text.  Not an option.

File: lib/MooX/Options/Descriptive/Usage.pm (hunk 1 above)
Comment (kablamo) [inline lib/MooX/Options/Descriptive/Usage.pm]: Thats something Getopt::Long::Descriptive does to indicate this $opt is not an option -- its whitespace or text.

File: lib/MooX/Options/Role.pm (hunk 2 above)
Comment (kablamo) [inline lib/MooX/Options/Role.pm]: This works.  But if you have a better solution I'm happy to hear it.  What kind of patch are you thinking of?

File: lib/MooX/Options/Descriptive/Usage.pm (hunk 1 above)
Comment (kablamo) [inline lib/MooX/Options/Descriptive/Usage.pm]: this line applies when $opt is text like it probably would be when someone used usage_top or usage_bottom

File: lib/MooX/Options/Role.pm (hunk 3 above)
Comment (kablamo) [inline lib/MooX/Options/Role.pm]: $opt is a Getopt::Long::Descriptive::Opts object.  The help attribute only exists if you pass a help option to Getopt::Long::Descriptive::describe_options().  

1. I'm not sure what attributes you want rexistanceemoved from which role?
2. I am checking the existence of the help method on the line after this one.  Perhaps I am misunderstanding something?

**Comment by celogeek** [comment]
Ok thanks for clarification, I will create 3 patches, 1 for spacer, 1 for usage_top/bottom, 1 for optional auto help/usage
Also I need to create the compact help message.
I'm working on that, I will use your patch. thanks

**Comment by celogeek** [comment]
I can't do "spacer" option like getoptlong descriptive, the option is sort using "order" then "name".

the best way to do it, is to add a "spacer_after" => 1, so the option that end a block can create a spacer after, or we can also add a "spacer_before" => 1, to do the same for the first arg of a block.

**Comment by kablamo** [comment]
Can you talk more about why you can't do spacer?  I don't understand.  I think I need to see an example or a pull request.  Order did not seem to be an issue the way I did it in this pr.  See the test case I added.

**Comment by celogeek** [comment]
As I say the order is not the order of the option you declare, it is sorted by "order" attribute, then by "name" of the attribute.

So placing a spacer at the right place is no possible.

But you can add a spacer_after and spacer_before an attribute and then kind of control where the space is.

Take my devel branch, check the spacer test, try it.

I support man page, long usage, and also short usage !

**Comment by kablamo** [comment]
Thanks!  I checked out your devel branch.  I think you are misunderstanding spacer.  Checkout these Getopt::Long::Descriptive examples:

**Example 1**
```
my ($opt, $usage) = describe_options(
    '%c %o',
    [ 'help|h', 'Prints this message' ],
);  
```
Produces this usage:
```
myprogram.pl [-h] [long options...]
        -h --help    Prints this message
```

**Example 2**
```
my ($opt, $usage) = describe_options(
    '%c %o',
    [],
    [ 'help|h', 'Prints this message' ],
    [],
);  
```
Produces this usage (note the added newlines):
```
myprogram.pl [-h] [long options...]

        -h --help    Prints this message

```

**Example 3**
```
my ($opt, $usage) = describe_options(
    '%c %o',
    [],
    ['This script downloads front page Hacker News articles.'],
    [],
    [ 'debug|d',   'Debug output' ],
    [ 'verbose|v', 'Verbose output' ],
    [ 'help|h',    'Prints this message' ],
    [],
);  
```
Produces this usage:
```
myprogram.pl [-dhv] [long options...]
                      
        This script downloads front page Hacker News articles.
                      
        -d --debug      Debug output
        -v --verbose    Verbose output
        -h --help       Prints this message
                      
```

In this last example, Getopt::Long::Descriptive lets you know that `[]` and `['This script downloads front page Hacker News articles.']` are not options by setting `$opt->{desc} = 'spacer'`.

**Comment by celogeek** [comment]
Yes, I understand spacer, this separate 2 options. What you have to know is that I have rewrite the "Descriptive::Usage". That mean this is not handle exactly as the original version.
It support man, short also, and display man in a more expansive way.

I have set the space to "_" by default, but it can be replaced by a " " if you prefer that.

if I take your example 3 without the text message, it can be done this way :

```
use Moo;
use MooX::Options spacer => " ";

option debug => (is => 'ro', order => 0, spacer_before => 1);
option verbose => (is => 'ro', order => 1);
```

debug and version are already in the right order of course.
But if you want to group option and ensure one of them is on top, you can do this :

```
use Moo;
use MooX::Options spacer => " ";

option group1First => (is => 'ro', order => 0, spacer_before => 1);
option group1B => (is => 'ro', order => 1);
option group1C => (is => 'ro', order => 1);
option group1D => (is => 'ro', order => 1);
option group1E => (is => 'ro', order => 1);

option group2First => (is => 'ro', order => 2, spacer_before => 1);
option group2B => (is => 'ro', order => 3);
option group2C => (is => 'ro', order => 3);
option group2D => (is => 'ro', order => 3);
```

You can this way with "order" ensure than an option is place first of his group, and then add a spacer_before. Same can be done for the last.

About the free text, it isn't better to place the leading and post text before and after the usage ? like this :

```
This script downloads front page Hacker News articles.

myprogram.pl [-dhv] [long options...]

        -d --debug      Debug output
        -v --verbose    Verbose output
        -h --help       Prints this message

Design by me, 2015
```

So we can more easily add a text before and after the classic usage ?

Also about the "group", may be we can add a "group" option, with a text for it like this :

```
use Moo;
use MooX::Options spacer => " ";

option_group(1, 'This is my first group', spacer_before => 1);
option_group(2, 'This is my second group', spacer_before => 2);

option group1A => (is => 'ro', group => 1);
option group1B => (is => 'ro', group => 1);
option group1C => (is => 'ro', group => 1);
option group1D => (is => 'ro', group => 1);
option group1E => (is => 'ro', group => 1);

option group2A => (is => 'ro', group => 2);
option group2B => (is => 'ro', group => 2);
option group2C => (is => 'ro', group => 2);
option group2D => (is => 'ro', group => 2);
option group2E => (is => 'ro', group => 2);

```

This way it is easy to group option and describe them ?

**Comment by kablamo** [comment]
Thanks for the explanation.  That helps.  

About defaults:
- I rarely see "_" as a spacer in unix commands so it seems like a non standard default.  
- I don't think I ever see free text before the usage string in unix commands so it also seems a like a non standard default. 

`git help`, `ls --help`, `cp --help`, `rackup -h`, `knife help` all have the usage string first and all of them use " " as a spacer instead of "_".

spacer_before and spacer_after are good ideas I think.  option_group is probably a good idea too.  

However none of these features solve the problem I am trying to solve with this pr.  I want to add free text after the usage string and at the very end (usage_top, usage_bottom options).  How can I do that?

**Comment by celogeek** [comment]
First think I see with the actual version may be an "around" option_usage
You should be able to add pre and post text.

I will continue as soon as possible, I write the idea, I will try to fix the default with spaces.

**Comment by celogeek** [comment]
I have apply fixes and improvement on devel branch, can you check ?
I have change the spacer to " " by default.
Also do a more compact help message by default

**Comment by cazador481** [comment]
@celogeek Any idea when these changes are going to be pushed to master?  I like the new compact help.  Though I do request reducing the number of blank lines in the long help message.

**Comment by kablamo** [comment]
I looked at the devel branch.  But there are extra newlines:

```

USAGE: spacer.t [-abch] [long options...]

    -a       foo

    -b       foo

    -c       foo

    --usage  show a short help message
    -h       show a compact help message
    --help   show a long help message
    --man    show the manual
```

It should look like this instead:
```
USAGE: spacer.t [-abch] [long options...]

    -a       foo
    -b       foo
    -c       foo
    --usage  show a short help message
    -h       show a compact help message
    --help   show a long help message
    --man    show the manual
```

**Comment by celogeek** [comment]
Hi,

Yeah, lot's of project, I need to take this more seriously. I will try to complete and release it as soon as possible. Hopefully this week.

Thanks for comments, I was waiting for it ;)

I will reduce the space.

**Comment by kablamo** [comment]
awesome!

**Comment by cazador481** [comment]
Just as an Fyi I use this all the time.  I really love this module.

**Comment by celogeek** [comment]
I have check devel :

```
perl -Ilib -MMoo -MMooX::Options -E 'option "a" => (is => "ro", required => 1, doc => "aaaa"); option "b" => (is => "ro", doc => "bbbb"); option "c" => (is => "ro", long_doc => "cccc"); local @ARGV=qw(-h); main->new_with_options'
USAGE: -e [-abch] [long options...]

    -a       aaaa
    -b       bbbb
    -c       no doc for c
                                                                        
    --usage  show a short help message
    -h       show a compact help message
    --help   show a long help message
    --man    show the manual

```

Why do you have a different output ?

**Comment by celogeek** [comment]
I have release the v4.019, still need to do more stuff, but at least we got spacer and shorter help message by default.
I continue the other task

**Comment by kablamo** [comment]
\0/ :balloon:

**Comment by rehsack** [comment]
Can you please rebase and upload again?

**Pull request closed without merging**